django/db/migrations/migration.py

```diff
@@ -139,18 +175,37 @@ def unapply(self, project_state, schema_editor, collect_sql=False):
         """
         # Construct all the intermediate states we need for a reverse migration
         to_run = []
-        new_state = project_state
+
         # Phase 1
-        for operation in self.operations:
-            # If it's irreversible, error out
-            if not operation.reversible:
-                raise IrreversibleError("Operation %s in %s is not reversible" % (operation, self))
-            # Preserve new state from previous run to not tamper the same state
-            # over all operations
-            new_state = new_state.clone()
-            old_state = new_state.clone()
-            operation.state_forwards(self.app_label, new_state)
-            to_run.insert(0, (operation, old_state, new_state))
+        def _collect_operations(operations, to_state):
+            if not operations:
+                return
+
+            for operation in operations:
+                # If it's irreversible, error out
+                if not operation.reversible:
+                    raise IrreversibleError("Operation %s in %s is not reversible" % (operation, self))
+                # Preserve new state from previous run to not tamper the same state
+                # over all operations
+                to_state = to_state.clone()
+                from_state = to_state.clone()
+                operation.state_forwards(self.app_label, to_state)
+                to_run.insert(0, (operation, from_state, to_state))
+
+                # Insert injected operations from post_operation signal receivers.
+                injected_operations = emit_post_operation_signal(
+                    migration=self,
+                    operation=operation,
+                    from_state=from_state.clone(),
+                    to_state=to_state.clone(),
+                )
+
+                _collect_operations(injected_operations, to_state)
+
+        try:
+            _collect_operations(self.operations, project_state)
```

Comment:
Any reason not to call this one `_unapply_operations` to be coherent with `_apply_operations` in `apply`?

Reply:
I called it `collect` because unapplying happens in 2 phases. In the first one, we "collect" the list of operations by going forward and inserting in reverse order, and in phase two we perform `operation.database_backwards`. I was just concerned that using "unapply" would be misleading, but if you think it's more coherent, then it's fine with me. WDYT?

Reply:
I went for your suggestion as it actually reads pretty well in the code.